Accept passthrough parameters

slifty opened the pull request:
This PR modifies mw2pdf to accept `passthroughParameters` and decorate URLs with the specified data.  This makes certain authentication / configuration flows possible when book printing.

Issue #9

slifty committed:
Accept passthrough parameters

mw2pdf now takes a new `passthroughParameters` field which is
essentially an object that should be converted to a queryString and
appended to each URL.

This allows scripts to pass data back to a MediaWiki instance via
querystring.

Issue #9

hawkinsw commented on services/api/mw2pdf/index.js:
```diff
@@ -26,7 +27,15 @@ program
+    // Process passthrough parameters
+    let processedUrls = urls
+    if (options.passthroughParameters) {
+      const passthroughParameters = JSON.parse(options.passthroughParameters)
```
Do we want to handle the case where the user gives us malformed JSON?

slifty replied:
Crashing and burning is *kinda* like handling....

Great idea -- will add that on Monday.

slifty replied:
Thinking about this more: I actually do think that if malformed JSON is provided it should halt.

That said, I'm adding more explicit validation in the hopes that it will produce more friendly error messages..

hawkinsw replied:
This is great. As we discussed in the Zulip, we both agree that crashing and burning is the right thing to do. Making it explicit will really help!

Thanks for taking my feedback seriously!
Will

frankduncan replied:
Hijacking this comment to note that error handling is a place that the whole torque system is quite inadequate.  Luckily, like this, we make the very front door a bit better.  In this case, the only person passing these parameters is the system administrator who set the system up.

This isn't to say we shouldn't add these kinds of error checking, we should!  But it's more to call out that there's technical debt in general around malformed input processing in the whole system.  I kind of track it here: https://github.com/OpenTechStrategies/torque/issues/27

hawkinsw reviewed (approved):
Looks good to me. Merge away, if you like; I asked one small question.

slifty committed:
Add Joi

Joi is a data validation library with lots of fun tools for checking
user inputs.  We will use it to be a bit more explicit in ensuring that
more complex parameters are properly defined.

slifty committed:
Add validation to passthroughParameters

Instead of passively failing if the passthroughParameters does not
contain a valid JSON string we can be explicit and validate using Joi.

This also makes sure that the passthroughParameters is an object with
primitive values for each key.  This puts the burden of encoding
complex parameters on the application instead of imposing JSON.

hawkinsw commented on services/api/mw2pdf/utils/validation.js:
```diff
@@ -0,0 +1,31 @@
+    Joi.string()
+      .custom(
+        (value, helpers) => {
+          JSON.parse(value)
```
So, if `JSON.parse` throws an error, then the this `custom` validator will throw an error, too, which will _bubble up_ and cause the assert to fail? I just want to make sure that I am understanding this correctly?

slifty replied:
That irony when the validation logic is not actually explicit 😂

Yes exactly -- I may add a comment explaining that's what's going on here

slifty replied:
I made the error catching flow more explicit (and took the opportunity to make the resulting message a bit more clear too)

hawkinsw commented on services/api/mw2pdf/utils/validation.js:
```diff
@@ -0,0 +1,31 @@
+    JSON.parse(passthroughParameters),
+    Joi.object()
+      .pattern(
+        /\w+/,
```
This specifies that we are going to run this _validator_ for every key that matches the /w+/ pattern?

slifty replied:
Correct -- I'll add a comment explaining that since it is pretty opaque

slifty replied:
I put it in a named variable instead which I think is more clear -- let me know if that's the case or if it's still opaque

hawkinsw commented on services/api/mw2pdf/utils/validation.js:
```diff
@@ -0,0 +1,31 @@
+        /\w+/,
+        Joi.any().custom(
+          (value, helpers) => {
+            if (typeof value === 'object' && value !== null) {
```
We expect every key's value to be an object and not null?

slifty replied:
I'll make this a named helper method -- yes.

hawkinsw commented on services/api/mw2pdf/utils/validation.js:
```diff
@@ -0,0 +1,31 @@
+        Joi.any().custom(
+          (value, helpers) => {
+            if (typeof value === 'object' && value !== null) {
+              return helpers.error('any.invalid');
```
And if the key's value is either not an object or is null, then we add an error for this validation?

slifty replied:
Decided to use the explicit `throw` variety of validation instead since I think it's a bit more obvious what's going on.

hawkinsw commented on services/api/mw2pdf/index.js:
```diff
@@ -30,10 +32,20 @@ program
+          (url) => updateUrlParameters(url, passthroughParameters),
+        )
+      } catch (e) {
+        if (e instanceof Joi.ValidationError) {
```
I wish that JS had conditional catch clauses :-D

slifty replied:
Agreed :(

[This does exist but I decided not to even try it](https://stackoverflow.com/questions/20483597/conditional-catch-clauses-browsers-support/42607272)

hawkinsw commented on services/api/mw2pdf/index.js:
```diff
@@ -30,10 +32,20 @@ program
-        (url) => updateUrlParameters(url, passthroughParameters),
-      )
+      try {
+        assertValidPassthroughParameters(options.passthroughParameters)
```
Would you be willing to add a comment here that says that this is the line that will cause the exception that we are handling? Just might make it easier for me when I come along later and try to understand the code after several weeks away.

slifty replied:
My only hesitation is that might become a stale comment -- in theory there are other things that might cause exceptions.

The use of the term `assert` is actually meant to convey that it can throw errors (the idea being that `assert` in node is a thing that will throw an error if it does not produce a truthy value.) but I understand if that isn't actually intuitive.

I'm going to look into tweaking this to make it clearer what's going on.

slifty replied:
I fixed this up by making two separate try / catch blocks.

The second one is just to be defensive.

hawkinsw reviewed (commented):
Joi is really, really neat!

slifty commented on services/api/mw2pdf/utils/validation.js:
```diff
@@ -0,0 +1,45 @@
+  return (typeof value === 'object' && value !== null)
+}
+
+const validJsonStringSchema = Joi.string()
```
Just a note that I think the most "clean" way to do this would be to actually have schemas in their own directory, with each validation schema defined in its own file.

So we could do things like...

```
import {
  validJsonStringSchema,
  passthroughParameterSchema,
} from './schemas'
```

The only reason I'm not doing that right now is that I would like to make that improvement after we move to TypeScript.  But @hawkinsw if you think I should just make it happen I'd be fine to take that step as part of this PR.

slifty commented:
I got the geen light in DM!
